File: TVTComment/Model/IPC/IPCMessageFactory.cs
namespace TVTComment.Model.IPC
{
    static class IPCMessageFactory
    {
        public static IPCMessage.IIPCMessage FromRawIPCMessage(RawIPCMessage rawmsg)
        {
            IPCMessage.IIPCMessage ret = rawmsg.MessageName switch
            {
                "Chat" => new IPCMessage.ChatIPCMessage(),
                "ChannelList" => new IPCMessage.ChannelListIPCMessage(),
                "ChannelSelect" => new IPCMessage.ChannelSelectIPCMessage(),
                "CurrentChannel" => new IPCMessage.CurrentChannelIPCMessage(),
                "Time" => new IPCMessage.TimeIPCMessage(),
                "Close" => new IPCMessage.CloseIPCMessage(),
                "SetChatOpacity" => new IPCMessage.SetChatOpacityIPCMessage(),
                "Command" => new IPCMessage.CommandIPCMessage(),
                _ => throw new IPCMessage.IPCMessageDecodeException("不明なMessageNameです: " + rawmsg.ToString()),
            };
            ret.Decode(rawmsg.Contents);
            return ret;
        }
    }
}

File: NicoJK/TVTComment/IPC/IPCMessageFactory.cpp
#include "stdafx.h"
#include "IPCMessageFactory.h"
#include "IPCMessageDecodeError.h"
#include "IPCMessage/ChatIPCMessage.h"
#include "IPCMessage/ChannelListIPCMessage.h"
#include "IPCMessage/CurrentChannelIPCMessage.h"
#include "IPCMessage/ChannelSelectIPCMessage.h"
#include "IPCMessage/TimeIPCMessage.h"
#include "IPCMessage/CloseIPCMessage.h"
#include "IPCMessage/SetChatOpacityIPCMessage.h"
#include "IPCMessage/CommandIPCMessage.h"

namespace TVTComment
{
	std::unique_ptr<IIPCMessage> MakeIPCMessageFromRaw(const RawIPCMessage &rawmsg)
	{
		IIPCMessage *msg;
		if (rawmsg.MessageName == "Chat")
			msg = new ChatIPCMessage();
		else if (rawmsg.MessageName == "ChannelList")
			msg = new ChannelListIPCMessage();
		else if (rawmsg.MessageName == "CurrentChannel")
			msg = new CurrentChannelIPCMessage();
		else if (rawmsg.MessageName == "ChannelSelect")
			msg = new ChannelSelectIPCMessage();
		else if (rawmsg.MessageName == "Time")
			msg = new TimeIPCMessage();
		else if (rawmsg.MessageName == "Close")
			msg = new CloseIPCMessage();
		else if (rawmsg.MessageName == "SetChatOpacity")
			msg = new SetChatOpacityIPCMessage();
		else if (rawmsg.MessageName == "Command")
			msg = new CommandIPCMessage();
		else
			throw IPCMessageDecodeError("不明なMessageNameです: " + rawmsg.ToString());

		std::unique_ptr<IIPCMessage> ret(msg);
		ret->Decode(rawmsg.Contents);
		return ret;
	}
}